### PR TITLE
Complete refactoring

### PR DESCRIPTION
**Pull request by polomani**

### CODE REVIEW

File: src/__tests__/cogniteClient.integration.ts [hunk 1]
```diff
@@ -25,7 +28,7 @@ describe('http methods - integration', () => {
   });
   test('post method', async () => {
     const assets = [{ name: 'First asset' }, { name: 'Second asset' }];
-    const response = await client.post(
+    const response = await client.post<ItemsResponse<Asset[]>>(
```
Comment (polomani) [inline src/__tests__/cogniteClient.integration.ts]: default to any?

File: src/__tests__/cogniteClient.spec.ts [hunk 2]
```diff
@@ -105,39 +110,46 @@ describe('CogniteClient', () => {
...
   describe('http requests', () => {
-    let mock: MockAdapter;
     let client: CogniteClient;
 
     beforeAll(async () => {
-      client = setupClient();
-      mock = new MockAdapter(client.instance);
-    });
-    beforeEach(async () => {
-      mock.reset();
```
Comment (polomani) [inline src/__tests__/cogniteClient.spec.ts:116]: maybe `nock.reset()` or something?

File: src/__tests__/resources/assets.integration.spec.ts [hunk 3]
```diff
@@ -172,6 +173,18 @@ describe('Asset integration test', () => {
     expect(response.nextCursor).toBeDefined();
     expect(response.items).toBeDefined();
     expect(response.items[0].id).toBeDefined();
+    expect(response.items[0]).toBeInstanceOf(AssetClass);
+    // TODO: uncomment line below
```
Comment (polomani) [inline src/__tests__/resources/assets.integration.spec.ts]: uncomment

File: src/__tests__/resources/assets.integration.spec.ts [hunk 4]
```diff
@@ -172,6 +173,18 @@ describe('Asset integration test', () => {
...
+    // expect(response.items).toBeInstanceOf(AssetList);
+  });
+
+  test('list.next', async () => {
+    const response = await client.assets.list({ limit: 1 });
+    expect(response.next).toBeDefined();
+    const nextPage = await response.next!();
+    expect(nextPage.items[0]).toBeInstanceOf(AssetClass);
+    // TODO: uncomment line below
```
Comment (polomani) [inline src/__tests__/resources/assets.integration.spec.ts]: uncomment

File: src/__tests__/standardMethods.spec.ts [hunk 5]
```diff
@@ -1,220 +0,0 @@
...
-      axiosMock.onGet('/').reply(numberGeneratorAPIEndpoint);
-      return listEndpoint;
-    }
-    test('for-await', async () => {
-      const listEndpoint = createMockedListEndpoint();
-      const numbers = [];
-      for await (const item of listEndpoint()) {
-        numbers.push(item);
-      }
-      expect(numbers).toMatchSnapshot();
```
Comment (polomani) [inline src/__tests__/standardMethods.spec.ts:139]: do we have tests from this file somewhere?

File: src/__tests__/utils/http/basicHttpClient.spec.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        .reply(200, { a: 42 });
+      const response = await client.get<any>('/');
+      expect(response.data).toEqual({ a: 42 });
+    });
+
+    test('handle text response', async () => {
+      nock(baseUrl)
+        .get('/')
+        .reply(200, 'hello');
+      const response = await client.get<string>('/', {
```
Comment (polomani) [inline src/__tests__/utils/http/basicHttpClient.spec.ts:64]: maybe add arraybuffer test here as well?

File: src/__tests__/utils/http/cdfHttpClient.spec.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+      test('set custom 401 handler', async done => {
+        nock(baseUrl)
+          .get('/')
+          .reply(401, error401);
+        client.set401ResponseHandler(err => {
+          expect(err.status).toBe(401);
+          done();
+        });
+        client.get('/');
```
Comment (polomani) [inline src/__tests__/utils/http/cdfHttpClient.spec.ts:133]: await?

File: src/__tests__/utils/http/cdfRetryValidator.spec.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    ...baseResponse,
+    status: 500,
+  };
+
+  test("don't retry 2xx", () => {
+    expect(cdfRetryValidator(getRequest, response200, 0)).toBeFalsy();
+  });
+
+  test('only retry reasonable times', () => {
+    expect(cdfRetryValidator(getRequest, response500, 1000)).toBeFalsy();
```
Comment (polomani) [inline src/__tests__/utils/http/cdfRetryValidator.spec.ts]: lower the number here? to 10?

File: src/__tests__/utils/http/cdfHttpClient.spec.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+          .get('/')
+          .reply(401, error401);
+        nock(baseUrl)
+          .get('/')
+          .reply(200, {});
+        client.set401ResponseHandler((_, retry) => {
+          retry();
+        });
+        expect((await client.get('/')).status).toBe(200);
+        expect(scope.isDone).toBeTruthy();
```
Comment (polomani) [inline src/__tests__/utils/http/cdfHttpClient.spec.ts]:
```suggestion
        expect(scope.isDone()).toBeTruthy();
```

File: src/__tests__/resources/revisions3D.integration.spec.ts [hunk 10]
```diff
@@ -62,7 +62,10 @@ describeIfCondition(
         ]),
         // upload file
         client.files.upload(
-          { name: `file_revision_test_${randomInt()}.fbx` },
+          {
+            name: `file_revision_test_${randomInt()}.fbx`,
+            mimeType: 'application/octet-stream',
```
Comment (ghost) [inline src/__tests__/resources/revisions3D.integration.spec.ts:67]: Breaking change? Or can we still upload the file without the `mimeType`

File: src/autoPagination.ts [hunk 11]
```diff
@@ -109,3 +109,8 @@ function makeAutoPagingToArray<T>(autoPagingEach: AutoPagingEach<T>) {
     return items;
   };
 }
+
+export type CursorAndAsyncIterator<T, Wrapper = T[]> = Promise<
```
Comment (ghost) [inline src/autoPagination.ts:113]: This is moved from `standardMethods.ts`. Breaking change?

File: src/cogniteClient.ts [hunk 12]
```diff
@@ -316,8 +317,8 @@ export default class CogniteClient {
    * const response = await client.get('/api/v1/projects/{project}/assets', { params: { limit: 50 }});
    * ```
    */
-  public get = (path: string, options?: BaseRequestOptions) =>
-    this.doRawRequest('get', path, options);
+  public get = <T = any>(path: string, options?: HttpRequestOptions) =>
```
Comment (ghost) [inline src/cogniteClient.ts:320]: Breaking change: BaseRequestOptions is different from HttpRequestOptions. (enum vs string)

File: src/cogniteClient.ts [hunk 13]
```diff
@@ -406,22 +422,3 @@ function onAuthenticateWithPopup(login: OnAuthenticateLoginObject) {
     redirectUrl: window.location.href,
   });
 }
-export interface BaseRequestOptions {
-  data?: any;
-  params?: object;
-  headers?: { [key: string]: string };
-  responseType?: 'json' | 'arraybuffer' | 'text';
-}
-export interface Response {
```
Comment (polomani) [inline src/cogniteClient.ts:415]: are those moved somewhere?

File: src/resources/3d/files3DApi.ts [hunk 14]
```diff
@@ -1,40 +1,22 @@
...
-  /** @hidden */
-  constructor(project: string, instance: AxiosInstance, map: MetadataMap) {
-    this.path = projectUrl(project) + '/3d/files';
-    this.instance = instance;
-    this.map = map;
-  }
+import { HttpResponseType } from '../../utils/http/basicHttpClient';
+import { BaseResourceAPI } from '../baseResourceApi';
 
+export class Files3DAPI extends BaseResourceAPI<any> {
```
Comment (polomani) [inline src/resources/3d/files3DApi.ts:7]: any?

File: src/resources/3d/models3DApi.ts [hunk 15]
```diff
@@ -63,42 +73,7 @@ export class Models3DAPI {
...
-    this.retrieve = generateRetrieveSingleEndpoint(instance, path, map);
+  public async delete(ids: InternalId[]) {
+    return super.deleteEndpoint(ids);
   }
 }
-
-export type Models3DListEndpoint = (
-  scope?: Model3DListRequest
-) => CursorAndAsyncIterator<Model3D>;
-
```
Comment (polomani) [inline src/resources/3d/models3DApi.ts:91]: where to find this type now?

File: src/types/types.ts [hunk 16]
```diff
@@ -166,6 +166,10 @@ export interface AssetChangeByExternalId extends AssetPatch, ExternalId {}
 
 export interface AssetChangeById extends AssetPatch, InternalId {}
 
+export interface AssetDeleteParams {
```
Comment (ghost) [inline src/types/types.ts]: This interface has switched file. Breaking change?

File: src/types/types.ts [hunk 17]
```diff
@@ -829,7 +835,7 @@ export interface InternalId {
 }
 
 export interface ItemsResponse<T> {
-  items: T[];
+  items: T;
```
Comment (ghost) [inline src/types/types.ts:837]: Breaking change?

File: src/resources/3d/revisions3DApi.ts [hunk 18]
```diff
@@ -167,77 +152,12 @@ export class Revisions3DAPI {
...
-  revisionId: CogniteInternalId,
-  query?: List3DNodesQuery
-) => CursorAndAsyncIterator<Node3D>;
-
-export type Revisions3DListNodesAncestorsEndpoint = (
-  modelId: CogniteInternalId,
-  revisionId: CogniteInternalId,
-  nodeId: CogniteInternalId,
-  query?: List3DNodesQuery
-) => CursorAndAsyncIterator<Node3D>;
```
Comment (polomani) [inline src/resources/3d/revisions3DApi.ts:227]: those types, where are they now?

File: src/resources/apiKeys/apiKeysApi.ts [hunk 19]
```diff
@@ -42,27 +45,7 @@ export class ApiKeysAPI {
...
-    this.delete = generateDeleteEndpoint(instance, path, map);
+  public async delete(ids: CogniteInternalId[]): Promise<{}> {
+    return super.deleteEndpoint(ids);
   }
 }
-
-export type ApiKeysListEndpoint = (
-  scope?: ApiKeyListScope
-) => Promise<ApiKeyObject[]>;
-
```
Comment (polomani) [inline src/resources/apiKeys/apiKeysApi.ts:63]: types?

File: src/resources/assets/assetsApi.ts [hunk 20]
```diff
@@ -190,32 +166,3 @@ export class AssetsAPI {
...
-
-export type AssetListEndpoint = (
-  scope?: AssetListScope
-) => CursorAndAsyncIterator<Asset>;
-
-export type AssetRetrieveEndpoint = (
-  ids: AssetIdEither[]
-) => Promise<AssetList>;
-
-export type AssetUpdateEndpoint = (
```
Comment (polomani) [inline src/resources/assets/assetsApi.ts:206]: types removed?

File: src/resources/dataPoints/dataPointsApi.ts [hunk 21]
```diff
@@ -35,7 +31,11 @@ export class DataPointsAPI {
    * const data = await client.datapoints.retrieve({ items: [{ id: 123 }] });
    * ```
    */
-  public retrieve: DataPointsRetrieveEndpoint;
+  public async retrieve(
+    query: DatapointsMultiQuery
+  ): Promise<(DatapointsGetAggregateDatapoint | DatapointsGetDatapoint)[]> {
```
Comment (polomani) [inline src/resources/dataPoints/dataPointsApi.ts]: you sure it's not `DatapointsGetAggregateDatapoint[] | DatapointsGetDatapoint[]`?

File: src/resources/dataPoints/dataPointsApi.ts [hunk 22]
```diff
@@ -62,30 +66,48 @@ export class DataPointsAPI {
...
+      ItemsResponse<
+        (DatapointsGetAggregateDatapoint | DatapointsGetDatapoint)[]
+      >
+    >(path, {
+      data: query,
+    });
+    return this.addToMapAndReturn(response.data.items, response);
+  }
 
-export type DataPointsRetrieveLatestEndpoint = (
```
Comment (polomani) [inline src/resources/dataPoints/dataPointsApi.ts:85]: types

File: src/resources/events/eventsApi.ts [hunk 23]
```diff
@@ -87,38 +89,7 @@ export class EventsAPI {
...
-
-export type EventsListEndpoint = (
-  scope?: EventFilterRequest
-) => CursorAndAsyncIterator<CogniteEvent>;
-
-export type EventsRetrieveEndpoint = (
-  ids: IdEither[]
-) => Promise<CogniteEvent[]>;
-
-export type EventsUpdateEndpoint = (
```
Comment (polomani) [inline src/resources/events/eventsApi.ts:116]: removed types

File: src/resources/files/filesApi.ts [hunk 24]
```diff
@@ -95,62 +124,76 @@ export class FilesAPI {
...
+        return fileInfo;
+      }
+      retryCount++;
+      await sleepPromise(DELAY_IN_MS);
+    } while (retryCount < MAX_RETRIES);
+    throw Error(`File never marked as 'uploaded'`);
+  }
 
-export type FilesUpdateEndpoint = (
-  changes: FileChangeUpdate[]
```
Comment (polomani) [inline src/resources/files/filesApi.ts:155]: removed types

File: src/resources/groups/groupsApi.ts [hunk 25]
```diff
@@ -116,42 +98,17 @@ export class GroupsAPI {
...
+  private encodeServiceAccountUrl = (groupId: number) =>
+    this.url(`${groupId}/serviceaccounts`);
 }
-
-export type GroupsListEndpoint = (scope?: ListGroups) => Promise<Group[]>;
-
-export type GroupsCreateEndpoint = (items: GroupSpec[]) => Promise<Group[]>;
-
-export type GroupsDeleteEndpoint = (ids: CogniteInternalId[]) => Promise<{}>;
-
```
Comment (polomani) [inline src/resources/groups/groupsApi.ts:144]: types?

File: src/resources/login/loginApi.ts [hunk 26]
```diff
@@ -19,7 +19,7 @@ export class LoginAPI {
    * // if status === null means you are not logged in
    * ```
    */
-  public status: LoginStatusEndpoint = () => getIdInfo(this.instance, {});
+  public async status(): Promise<null | IdInfo> {
+    return getIdInfo(this.httpClient, {});
+  }
 }
-
-export type LoginStatusEndpoint = () => Promise<null | IdInfo>;
```
Comment (polomani) [inline src/resources/login/loginApi.ts:25]: type?

File: src/resources/projects/projectsApi.ts [hunk 27]
```diff
@@ -39,20 +27,18 @@ export class ProjectsAPI {
...
+  ): Promise<ProjectResponse> {
+    const path = this.encodeUrl(projectName);
+    const response = await this.httpClient.put<ProjectResponse>(path, {
+      data: replacement,
+    });
+    return this.addToMapAndReturn(response.data, response);
+  }
 
-export type ProjectsUpdateEndpoint = (
-  urlName: string,
```
Comment (polomani) [inline src/resources/projects/projectsApi.ts:56]: type?

File: src/resources/raw/rawApi.ts [hunk 28]
```diff
@@ -199,78 +162,15 @@ export class RawAPI {
...
-export type RawCreateDatabasesEndpoint = (items: RawDB[]) => Promise<RawDB[]>;
-
 export interface RawDatabaseDeleteParams {
   recursive?: boolean;
 }
-export type RawDeleteDatabasesEndpoint = (
-  items: RawDB[],
-  params?: RawDatabaseDeleteParams
-) => Promise<{}>;
-
```
Comment (polomani) [inline src/resources/raw/rawApi.ts:234]: types?

File: src/resources/raw/rawTablesApi.ts [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+    );
+  }
+
+  public async delete(databaseName: string, items: RawDBTable[]) {
+    const path = `${this.encodeUrl(databaseName)}/tables/delete`;
+    await this.postInParallelWithAutomaticChunking(path, items);
+    return {};
+  }
+
+  protected transformToList(items: RawDB[]) {
```
Comment (polomani) [inline src/resources/raw/rawTablesApi.ts]: didn't we have it as default now?

File: src/resources/raw/rawTablesApi.ts [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+    return items;
+  }
+
+  protected transformToClass(items: RawDB[]) {
+    return items;
```
Comment (polomani) [inline src/resources/raw/rawTablesApi.ts]: and this?

File: src/resources/securityCategories/securityCategoriesApi.ts [hunk 31]
```diff
@@ -46,25 +47,7 @@ export class SecurityCategoriesAPI {
...
-    this.delete = generateDeleteEndpoint(instance, path, map);
+  public async delete(ids: CogniteInternalId[]): Promise<{}> {
+    return super.deleteEndpoint(ids);
   }
 }
-
-export type SecurityCategoriesListEndpoint = (
-  query?: ListSecurityCategories
-) => CursorAndAsyncIterator<SecurityCategory>;
-
```
Comment (polomani) [inline src/resources/securityCategories/securityCategoriesApi.ts:63]: types?

File: src/resources/serviceAccounts/serviceAccountsApi.ts [hunk 32]
```diff
@@ -44,23 +46,7 @@ export class ServiceAccountsAPI {
...
-    this.create = generateCreateEndpoint(instance, path, map);
-    this.delete = generateDeleteEndpoint(instance, path, map);
+  public async delete(ids: CogniteInternalId[]): Promise<{}> {
+    return super.deleteEndpoint(ids);
   }
 }
-
-export type ServiceAccountsListEndpoint = () => Promise<ServiceAccount[]>;
-
-export type ServiceAccountsCreateEndpoint = (
```
Comment (polomani) [inline src/resources/serviceAccounts/serviceAccountsApi.ts:60]: types?

File: src/resources/timeSeries/timeSeriesApi.ts [hunk 33]
```diff
@@ -99,70 +117,16 @@ export class TimeSeriesAPI {
...
-) => Promise<TimeSeriesList>;
-
-export type TimeSeriesSearchEndpoint = (
-  query: TimeSeriesSearchDTO
-) => Promise<TimeSeriesList>;
-
-export type TimeSeriesRetrieveEndpoint = (
-  ids: TimeseriesIdEither[]
-) => Promise<TimeSeriesList>;
-
```
Comment (polomani) [inline src/resources/timeSeries/timeSeriesApi.ts:161]: types?

File: src/__tests__/resources/revisions3D.integration.spec.ts (hunk 10 above)
Comment (ghost) [reply src/__tests__/resources/revisions3D.integration.spec.ts:67]: I checked now and it works without specifying `mimeType`

File: src/resources/3d/viewer3DApi.ts [hunk 34]
```diff
@@ -121,55 +127,10 @@ export class Viewer3DAPI {
...
-  modelId: CogniteInternalId,
-  revisionId: CogniteInternalId
-) => Promise<RevealRevision3D>;
-
-export type Viewer3DListRevealNodes3DEndpoint = (
-  modelId: CogniteInternalId,
-  revisionId: CogniteInternalId,
-  query?: List3DNodesQuery
-) => CursorAndAsyncIterator<RevealNode3D>;
-
```
Comment (polomani) [inline src/resources/3d/viewer3DApi.ts:146]: types?

File: src/resources/3d/assetMappings3DApi.ts [hunk 35]
```diff
@@ -95,43 +79,16 @@ export class AssetMappings3DAPI {
...
-  filter?: AssetMappings3DListFilter
-) => CursorAndAsyncIterator<AssetMapping3D>;
-
-export type AssetMappings3DCreateEndpoint = (
-  modelId: CogniteInternalId,
-  revisionId: CogniteInternalId,
-  items: CreateAssetMapping3D[]
-) => Promise<AssetMapping3D[]>;
-
-export type AssetMappings3DDeleteEndpoint = (
```
Comment (polomani) [inline src/resources/3d/assetMappings3DApi.ts:123]: types?

File: src/__tests__/cogniteClient.integration.ts (hunk 1 above)
Comment (ghost) [reply src/__tests__/cogniteClient.integration.ts]: Done

File: src/cogniteClient.ts (hunk 12 above)
Comment (ghost) [reply src/cogniteClient.ts:320]: Done

**Review by ghost — APPROVED** [review]
LGTM 😎

**Pull request merged**